pkg/controller/certificates/certificate_controller.go

```diff
@@ -46,23 +46,39 @@ type CertificateController struct {
 
 	handler func(*certificates.CertificateSigningRequest) error
 
-	queue workqueue.RateLimitingInterface
+	queuingDelay time.Duration
+	queue        workqueue.RateLimitingInterface
 }
 
 func NewCertificateController(
 	kubeClient clientset.Interface,
 	csrInformer certificatesinformers.CertificateSigningRequestInformer,
 	handler func(*certificates.CertificateSigningRequest) error,
 ) (*CertificateController, error) {
+	return NewCertificateControllerWithDelay(
+		kubeClient,
+		csrInformer,
+		handler,
+		0,
+	)
+}
+
+func NewCertificateControllerWithDelay(
```

Comment:
Not that I'm aware of, unless an error is returned from the CSR processing. If an error is returned, it will be requeued for reprocessing.

Reply:
If I've understood the queue correctly, then [this method](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/certificate_controller.go#L130-L146) in the certificate controller results in only requeuing the item if processing it resulted in an error.

Reply:
So from what I understand the current code:

* Views a CSR.
* Adds the name of the CSR to the queue but waits 1 hour to process it.
* After 1 hour, CSR is processed, if it's been rejected it's removed.
* Regardless the CSR is removed from the work queue.

Since the queue is currently a "inspect CSR exactly 1 hour after every time it's modified" I'm less sure what happens if we need to inspect CSRs multiple times. e.g. if we start also removing un-acted upon CSRs after 24 hours.

Reply:
Adding criteria for removing CSRs after 24 hours (or any other time interval different from the existing interval) definitely requires a different approach.

I see 2 options.
1. Return an error if the CSR is not ready to be processed. It's not really an error if the enqueued items isn't ready yet, so this seems a little questionable.
2. Change the interface of the handler so includes a return value with a retry interval.

Reply:
Working on a third option, based on @liggitt's comment. Will just do a full resync at a polling interval.

Reply:
Done. Now polls once an hour and applies the different criteria for deleting old CSRs.